Fix InvalidLossStopping callback to be compatible with TDVP drivers (issue #1895)

lindamauron opened the pull request:
As mentioned in #1895, the loss is not evaluated before starting TDVP and calling callbacks.
 
Here, I simply added this estimation in the initialization of the driver. I also added a small test to verify that it works well and that the callback does not crash anymore.

PhilipVinc commented on netket/experimental/driver/tdvp_common.py:
```diff
@@ -99,6 +99,7 @@ def __init__(
             self._generator = lambda _: op
         else:
             self._generator = operator
+        self._loss_stats = self.state.expect(self.generator(t0))
```
That's not ok. 
This quantity will never be updated later so your callback will just see this value forever.

The generator value is already computed at every (sub)step, for example [here](https://github.com/netket/netket/blob/8bf6fbbdfc6e1e359c613a495a22e6d29f6a4806/netket/experimental/driver/tdvp.py#L137) but it's simply not stored in the driver itself.

You could store during the stage 0 (which corresponds to the current time, which is what you want) by setting it here https://github.com/netket/netket/blob/8bf6fbbdfc6e1e359c613a495a22e6d29f6a4806/netket/experimental/driver/tdvp.py#L151

PhilipVinc replied:
(Please fix also tdvp_schmitt in a similar place)

lindamauron replied:
Isn't it the value update just [here](https://github.com/netket/netket/blob/8bf6fbbdfc6e1e359c613a495a22e6d29f6a4806/netket/experimental/driver/tdvp.py#L139C12-L139C23) ? At least, it is the attribute used for the progress bar and the logging. And the value (of <h>) reported in the log changes, so it seems to be updated. 

Or maybe there is something I am missing...

lindamauron replied:
Indeed, if we want to avoid supplementary computations, it seems to be the best option

lindamauron commented:
I changed as you suggested, leaving the TDVP drivers as before but only impacting the callback. I had to adjust it a bit so it works well with TDVP, but its behavior with VMC is unchanged (I also added tests with both vmc and tdvp)

PhilipVinc commented on netket/callbacks/invalidlossstopping.py:
```diff
@@ -28,8 +28,8 @@ class InvalidLossStopping(struct.Pytree, mutable=True):
     # caches
-    _last_valid_iter: int
-    """Last valid iteration, to check against patience"""
+    _invalid_steps: int
```
Can you explain this change? 
The reason why we were storing the 'last valid iter' instead of 'invalid step' was because sometimes we do not call the callback at every step (if you specify step_size=2 in run for example).

So we need to store the interval..

PhilipVinc replied:
Ah, I think I understand. 
Is it because `step` was a float that is the time? and you wanted an integer?

Then you can use https://netket.readthedocs.io/en/latest/api/_generated/driver/netket.driver.AbstractVariationalDriver.html#netket.driver.AbstractVariationalDriver.step_count 

instead of step

PhilipVinc replied:
yep, that's why tdvp driver has both `driver.step_count` and `driver.step_value` . You can safely ignore the `step` argument passed to the callback (it will be identical to `driver.step_value` IIRC, and you can instead use `driver.step_count` which is an integer counter.

I was suggesting above to you step count because this will allow to not change the behaviour for VMC & similar optimisers, while you get what you want for tdvp (I think)
